src/modules/musig/keyagg_impl.h

```diff
@@ -0,0 +1,271 @@
+     * we test this) */
+    VERIFY_CHECK(ret);
+    secp256k1_musig_keyaggcoef_internal(sc, ctx->pk_hash, &pt->x, &ctx->second_pk_x);
+    return 1;
+}
+
+int secp256k1_musig_pubkey_agg(const secp256k1_context* ctx, secp256k1_scratch_space *scratch, secp256k1_xonly_pubkey *agg_pk, secp256k1_musig_keyagg_cache *keyagg_cache, const secp256k1_xonly_pubkey * const* pubkeys, size_t n_pubkeys) {
+    secp256k1_musig_pubkey_agg_ecmult_data ecmult_data;
+    secp256k1_gej pkj;
+    secp256k1_ge pkp;
+    size_t i;
+
+    VERIFY_CHECK(ctx != NULL);
+    if (agg_pk != NULL) {
+        memset(agg_pk, 0, sizeof(*agg_pk));
+    }
+    ARG_CHECK(pubkeys != NULL);
+    ARG_CHECK(n_pubkeys > 0);
+
+    ecmult_data.ctx = ctx;
+    ecmult_data.pks = pubkeys;
+    /* No point on the curve has an X coordinate equal to 0 */
+    secp256k1_fe_set_int(&ecmult_data.second_pk_x, 0);
+    for (i = 1; i < n_pubkeys; i++) {
+        secp256k1_ge pt;
+        if (!secp256k1_xonly_pubkey_load(ctx, &pt, pubkeys[i])) {
+            return 0;
+        }
+        if (secp256k1_memcmp_var(pubkeys[0], pubkeys[i], sizeof(*pubkeys[0])) != 0) {
+            ecmult_data.second_pk_x = pt.x;
+            break;
+        }
+    }
+
+    if (!secp256k1_musig_compute_pk_hash(ctx, ecmult_data.pk_hash, pubkeys, n_pubkeys)) {
+        return 0;
+    }
+    if (!secp256k1_ecmult_multi_var(&ctx->error_callback, scratch, &pkj, NULL, secp256k1_musig_pubkey_agg_callback, (void *) &ecmult_data, n_pubkeys)) {
+        /* The current implementation of ecmult_multi_var makes this code unreachable with tests. */
+        return 0;
+    }
+    secp256k1_ge_set_gej(&pkp, &pkj);
+    secp256k1_fe_normalize_var(&pkp.y);
+    /* The resulting public key is infinity with negligible probability */
+    VERIFY_CHECK(!secp256k1_ge_is_infinity(&pkp));
+    if (keyagg_cache != NULL) {
+        secp256k1_keyagg_cache_internal cache_i = { 0 };
+        cache_i.pk = pkp;
+        cache_i.second_pk_x = ecmult_data.second_pk_x;
+        memcpy(cache_i.pk_hash, ecmult_data.pk_hash, sizeof(cache_i.pk_hash));
+        secp256k1_keyagg_cache_save(keyagg_cache, &cache_i);
+    }
+
+    secp256k1_extrakeys_ge_even_y(&pkp);
+    if (agg_pk != NULL) {
+        secp256k1_xonly_pubkey_save(agg_pk, &pkp);
+    }
+    return 1;
+}
+
+int secp256k1_musig_pubkey_tweak_add(const secp256k1_context* ctx, secp256k1_pubkey *output_pubkey, const unsigned char *tweak32, secp256k1_musig_keyagg_cache *keyagg_cache) {
+    secp256k1_keyagg_cache_internal cache_i;
+    int overflow = 0;
+    secp256k1_scalar tweak;
+
+    VERIFY_CHECK(ctx != NULL);
+    if (output_pubkey != NULL) {
+        memset(output_pubkey, 0, sizeof(*output_pubkey));
+    }
+    ARG_CHECK(keyagg_cache != NULL);
+    ARG_CHECK(tweak32 != NULL);
+
+    if (!secp256k1_keyagg_cache_load(ctx, &cache_i, keyagg_cache)) {
+        return 0;
+    }
+    secp256k1_scalar_set_b32(&tweak, tweak32, &overflow);
+    if (overflow) {
```

Comment:
The documentation for this function says that `tweak32` must be valid according to `secp256k1_ec_seckey_verify`, but this function only checks for overflow and not zero (which ec_seckey_verify also checks).

Do you think it's worth also returning 0 when `tweak32` is the zero scalar (especially if this could help users detect errors)?

Reply:
I think this function should match the behavior of ec_pubkey_tweak_add and xonly_pubkey_add with respect to the tweak, who both require that the tweak does not overflow but allow the tweak to be zero. It's better to mention in the doc that the tweak must pass seckey_verify than that it shouldn't overflow, because only the former is actionable as we don't have a dedicated function for overflow.